engine/src/main/java/io/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java

```diff
@@ -110,6 +115,30 @@ public void applyState(final long elementInstanceKey, final ProcessInstanceRecor
     }
   }
 
+  private void cleanupSequenceFlowsTaken(final ProcessInstanceRecord value) {
+    if (value.getBpmnElementType() != BpmnElementType.PARALLEL_GATEWAY) {
+      return;
+    }
+
+    final var parallelGateway =
+        processState.getFlowElement(
+            value.getProcessDefinitionKey(), value.getElementIdBuffer(), ExecutableFlowNode.class);
+
+    final var tokensBySequenceFlow =
+        sequenceFlowAnalyzer.determineTakenIncomingFlows(value.getFlowScopeKey(), parallelGateway);
+
+    // cleanup deferred records for sequence flow taken to this parallel gateway
+    tokensBySequenceFlow.forEach(
+        (sequenceFlow, tokens) -> {
+          // before a parallel gateway is activated, each incoming sequence flow of the gateway must
+          // be taken (at least once). if a sequence flow is taken more than once then the redundant
+          // token remains for the next activation of the gateway (Tetris principle)
+          final var firstToken = tokens.get(0);
+          elementInstanceState.removeStoredRecord(
+              value.getFlowScopeKey(), firstToken.getKey(), Purpose.DEFERRED);
+        });
+  }
+
```

Comment:
Are deferred records are still used elsewhere? IF not then we can just say delete all. Maybe we rename these to awaiting parallel gw records or something. I think anyway we could simplify this, like I wrote above just storing the element id and some key for example.

Reply:
Seems to still be used in multi-instance body processor and event sub-process processor. Let's add an additional issue to follow-up on removing deferred records

Reply:
sounds good

Reply:
Created: https://github.com/camunda-cloud/zeebe/issues/6832